docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,227 @@
+# Client: layering and package structure
+
+## Desired layers
+
+### Transport
+
+The transport layer is concerned with round-tripping requests to an apiserver
+somewhere. It consumes a Config object with options appropriate for this.
+(That's most of the current client.Config structure.)
+
+Transport delivers an object that implements http's RoundTripper interface
+and/or can be used in place of http.DefaultClient to route requests.
+
+Transport objects are safe for concurrent use, and are cached and reused by
+subsequent layers.
+
+Tentative name: "Transport".
+
+It's expected that the transport config will be general enough that third
+parties (e.g., OpenShift) will not need their own implementation, rather they
+can change the Host field to point at their own servers, etc..
+
+Action items:
+* Split out of current client package into a new package. (@krousey)
+
+### RESTClient/request.go
+
+RESTClient consumes a Transport and a Codec (and optionally a group/version),
```

Comment:
I feel we use `Codec` a lot. Maybe we could clarify what this is? I think ideally it should be concerned with encoding and decoding only and not have anything to do with converting to different versions. Perhaps something like

```go
type Codec interface {
	Encode(runtime.Object) []byte
	Decode([]byte]) runtime.Object
}
```